clang/include/clang/AST/Type.h

```diff
@@ -3610,7 +3611,7 @@ class MemberPointerType : public Type, public llvm::FoldingSetNode {
   }
 
   void Profile(llvm::FoldingSetNodeID &ID) {
-    Profile(ID, getPointeeType(), getQualifier(), getMostRecentCXXRecordDecl());
+    Profile(ID, getPointeeType(), getQualifier(), getCXXRecordDecl());
```

Comment:
Is the problem that we run this profile in the 'middle' of a TU and the most recent decl changes by the end?  If so, this makes sense.

That said, I'd expect this to be the above suggestion in case this member pointer type has a redeclaration that perhaps changes what the actual referenced decl is.

```suggestion
    Profile(ID, getPointeeType(), getQualifier(), getCXXRecordDecl()->getCanonicalDecl());
```

Reply:
Yes, but this suggestion seems not needed because this `getCanonicalDecl()` call is already made inside the `Profile()` method: https://github.com/llvm/llvm-project/blob/e8ae77947154e10dbc05cbb95ec9e10d3b0be13e/clang/lib/AST/Type.cpp#L5279

Reply:
Hmm... then I'm surprised that this patch is needed at all for the purposes of non-determinism.  `getMostRecentCXXRecordDecl()->getCanonicalDecl()` will get the same declaration as `getCXXRecordDecl()->getCanonicalDecl()`.

Reply:
The issue is transient - the reproducibility rate on the reproducer (https://pastebin.com/6aL6rmBe) is only 1%. Also it's not the pointer we have here that's problematic, but rather the fact of extra AST node loading inside the `FoldingSet` hash calculation operation.

Reply:
> Hmm... then I'm surprised that this patch is needed at all for the purposes of non-determinism

My theory is that non-determinism is caused by side-effects in the deserialization logic that runs behind `getMostRecentCXXRecordDecl()`. We probably just get unlucky with rehashing points at different times for the hash table (`FoldingSet`) that makes us either call or not call `getMostRecentCXXRecordDecl`.

It is not a new issue, just rare and now getting surfaced by the new place where `getMostRecentCXXRecordDecl` is called.